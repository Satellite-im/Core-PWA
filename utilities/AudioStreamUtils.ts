import { AudioState } from '~/store/audio/types'

const IS_TALKING_SENSITIVITY = 10 // %
const TALKING_OFF_DELAY = 500 // ms

export class AudioStreamUtils {
  stream: MediaStream
  audio: AudioState | null = null

  requestId: number | null = null
  level = 0
  isTalking = false

  constructor(stream: MediaStream, audio?: AudioState) {
    this.stream = stream
    this.audio = audio || null
  }

  public start() {
    this.listenToMicLevel()
  }

  public destroy() {
    this.stopListeningToMicLevel()
  }

  private listenToMicLevel = () => {
    if (!this.stream?.active) {
      return
    }
    const audioContext = new AudioContext()
    const gainNode = audioContext.createGain()
    const analyser = audioContext.createAnalyser()
    const microphone = audioContext.createMediaStreamSource(this.stream)
    const javascriptNode = audioContext.createScriptProcessor(2048, 1, 1)

    analyser.smoothingTimeConstant = 0

    microphone.connect(gainNode)
    gainNode.connect(analyser)
    analyser.connect(javascriptNode)
    javascriptNode.connect(audioContext.destination)

    const array = new Uint8Array(analyser.frequencyBinCount)

    const updateState = () => {
      if (!this.audio) {
        console.error('No audio state')
        return
      }
      // Update gain based on inputVolume
      gainNode.gain.setValueAtTime(
        this.audio.inputVolume / 100,
        audioContext.currentTime,
      )
      this.requestId = requestAnimationFrame(updateState)

      analyser.getByteFrequencyData(array)
      let values = 0

      const length = array.length
      for (let i = 0; i < length; i++) {
        values += array[i]
      }

      this.level = values / length
      this.updateIsTalking(this.level)
    }

    updateState()
  }

  private stopListeningToMicLevel() {
    if (this.requestId) {
      cancelAnimationFrame(this.requestId)
      this.requestId = null
    }
  }

  private updateIsTalking(micLevel: number) {
    const talking = Math.round(micLevel) > IS_TALKING_SENSITIVITY
    let talkingOffBuffer

    if (!talking && this.isTalking) {
      talkingOffBuffer = setTimeout(() => {
        this.isTalking = false
      }, TALKING_OFF_DELAY)
    } else if (talking && !this.isTalking) {
      this.isTalking = true

      if (talkingOffBuffer) {
        clearTimeout(talkingOffBuffer)
      }
    }
  }
}
