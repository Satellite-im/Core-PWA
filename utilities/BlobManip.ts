/**
 * @function blobToStream
 * @description convert Blob to ReadableStream. helpful for large uploads to avoid hitting the browser blob size limit
 * @param {Blob} file
 * @returns {ReadableStream} base64 thumbnail
 */
export function blobToStream(file: Blob): ReadableStream {
  // @ts-ignore
  const reader = file.stream().getReader()
  const stream = new ReadableStream({
    start(controller) {
      function push() {
        return reader
          .read()
          .then(({ done, value }: { done: boolean; value: Uint8Array }) => {
            if (done) {
              controller.close()
              return
            }
            controller.enqueue(value)
            push()
          })
      }
      push()
    },
  })
  return stream
}

/**
 * @method blobToBase64
 * @description convert File/Blob to base64 string
 * @param {Blob} file
 * @returns {Promise<string>} base64 thumbnail
 */
export function blobToBase64(file: Blob): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    const reader = new FileReader()
    reader.readAsDataURL(file)
    reader.onload = () => resolve(reader.result?.toString() || '')
    reader.onerror = (error) => reject(error)
  })
}
