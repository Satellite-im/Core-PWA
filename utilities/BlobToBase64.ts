/**
 * @method blobToBase64
 * @description convert File/Blob to base64 string
 * @param {File} file
 * @returns {Promise<string>} base64 thumbnail
 */
export default function blobToBase64(file: File): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    const reader = new FileReader()
    reader.readAsDataURL(file)
    reader.onload = () => resolve(reader.result?.toString() || '')
    reader.onerror = (error) => reject(error)
  })
}
