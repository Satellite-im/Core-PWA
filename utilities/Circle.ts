export function addressToNumber(address: string) {
  return parseInt(address.slice(0, 10), 16)
}

export function toHex(str: string) {
  let res = ''
  for (let n = 0, l = str.length; n < l; n++) {
    const hex = Number(str.charCodeAt(n)).toString(16)
    res += hex
  }
  return res
}

export function stringToColor(str: string): string {
  const hash = stringToHash(str)
  return colors[hash]
}

function stringToHash(str: string): number {
  return (
    str.split('').reduce((p, c) => p + c.charCodeAt(0), 0) % (colors.length - 1)
  )
}

const colors = [
  '#fc5c65', // Fiery Rose
  '#fd9644', // Royal Orange
  '#fed330', // Sunglow
  '#26de81', // UFO Green
  '#2bcbba', // Maximum Blue Green
  '#eb3b5a', // Desire
  '#fa8231', // Princeton Orange
  '#f7b731', // Saffron
  '#20bf6b', // Green (Crayola)
  '#0fb9b1', // Tiffany Blue
  '#45aaf2', // Picton Blue
  '#4b7bec', // Blueberry
  '#a55eea', // Lavender Indigo
  '#778ca3', // Shadow Blue
  '#2d98da', // Bleu De France
  '#3867d6', // Royal Blue
  '#8854d0', // Amethyst
  '#a5b1c2', // Cadet Blue (Crayola)
  '#4b6584', // Dark Electric Blue
]
