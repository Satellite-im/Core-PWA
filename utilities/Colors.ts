export function hexToRGB(hex: string) {
  const int = parseInt(hex.replace('#', ''), 16)
  const r = (int >> 16) & 255
  const g = (int >> 8) & 255
  const b = int & 255
  return [r, g, b].join(', ')
}
