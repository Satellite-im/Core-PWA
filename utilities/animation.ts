export type AnimationConfig = {
  timing: (v: number) => number
  draw: (v: number) => void
  duration: number
}

export function easeOutBack(v: number): number {
  const c1 = 1.2
  const c3 = c1 + 1

  return 1 + c3 * Math.pow(v - 1, 5) + c1 * Math.pow(v - 1, 4)
}

export function animate(config: AnimationConfig) {
  const { timing, draw, duration } = config
  const start = performance.now()

  const animate = (time: number) => {
    let timeFraction = (time - start) / duration
    if (timeFraction > 1) {
      timeFraction = 1
    }

    const progress = timing(timeFraction)
    draw(progress)

    if (timeFraction < 1) {
      requestAnimationFrame(animate)
    }
  }
  requestAnimationFrame(animate)
}

export function lerp(x: number, y: number, a: number) {
  return x * (1 - a) + y * a
}
