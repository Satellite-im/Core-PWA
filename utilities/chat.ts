import { ConversationMessage } from '~/libraries/Iridium/chat/types'

export function conversationMessageIsNotice(message: ConversationMessage) {
  switch (message.type) {
    case 'member_join':
    case 'member_leave':
      return true
    default:
      return false
  }
}

export function isVisible(
  ele: HTMLElement,
  container: HTMLElement,
  partial = false,
) {
  const cTop = container.scrollTop
  const cBottom = cTop + container.clientHeight

  const eTop = ele.offsetTop
  const eBottom = eTop + ele.clientHeight

  const scrolledBefore = eTop >= cTop
  const scrolledAfter = eBottom <= cBottom
  const isTotal = scrolledBefore && scrolledAfter
  const isPartial =
    partial &&
    ((eTop < cTop && eBottom > cTop) || (eBottom > cBottom && eTop < cBottom))

  return { isElVisible: isTotal || isPartial, scrolledAfter, scrolledBefore }
}
