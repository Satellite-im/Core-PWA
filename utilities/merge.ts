export function overwriteMerge<T>(_a: any, b: T): T {
  return b
}
