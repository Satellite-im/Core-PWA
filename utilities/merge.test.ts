import * as merge from '~/utilities/merge'

describe('merge.overwriteMerge', () => {
  test('0', () => {
    const result: any = merge.overwriteMerge('#FF00FF', 'green')
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = merge.overwriteMerge('green', '#FF00FF')
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = merge.overwriteMerge(
      'rgb(0,100,200)',
      'rgb(0.1,0.2,0.3)',
    )
    expect(result).toMatchSnapshot()
  })

  test('3', () => {
    const result: any = merge.overwriteMerge('green', 'rgb(0.1,0.2,0.3)')
    expect(result).toMatchSnapshot()
  })

  test('4', () => {
    const result: any = merge.overwriteMerge('rgb(0.1,0.2,0.3)', '#FF00FF')
    expect(result).toMatchSnapshot()
  })

  test('5', () => {
    const result: any = merge.overwriteMerge('', '')
    expect(result).toMatchSnapshot()
  })
})
