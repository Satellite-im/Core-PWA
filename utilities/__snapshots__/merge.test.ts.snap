// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge.overwriteMerge 0 1`] = `"green"`;

exports[`merge.overwriteMerge 1 1`] = `"#FF00FF"`;

exports[`merge.overwriteMerge 2 1`] = `"rgb(0.1,0.2,0.3)"`;

exports[`merge.overwriteMerge 3 1`] = `"rgb(0.1,0.2,0.3)"`;

exports[`merge.overwriteMerge 4 1`] = `"#FF00FF"`;

exports[`merge.overwriteMerge 5 1`] = `""`;
